apps/src/lib/levelbuilder/code-docs-editor/ProgrammingEnvironmentEditor.jsx

```diff
@@ -55,13 +86,18 @@ export default function ProgrammingEnvironmentEditor({
         if (response.ok) {
           setLastUpdated(Date.now());
           setError(null);
+          return response.json();
```

Comment:
Can you explain to me what the changes here in the response to this fetch are doing?

Reply:
Yes! Because categories are their own model on the server, ones that already exist have their own unique identifier that we can use with react (`key`). However, when a new one is created, we assign it a random key (this is actually done in [OrderableList](https://github.com/code-dot-org/code-dot-org/blob/a5bf53d3a8db9b78bf79b4407ce502c7fdebd1cb/apps/src/lib/levelbuilder/code-docs-editor/OrderableList.jsx#L13)). When save is pressed, we want the page to use the new keys that the server has generated and sent back down, or else we can end up with duplicate categories. I copied this strategy off of what we do in [LessonEditor](https://github.com/code-dot-org/code-dot-org/blob/a5bf53d3a8db9b78bf79b4407ce502c7fdebd1cb/apps/src/lib/levelbuilder/lesson-editor/LessonEditor.jsx#L127) because that also saves multiple models on save. It looks a bit different here but `setProgrammingEnvironment` does pretty much the same this as what's going on in LessonEditor.

The actual `.json()` gets the return data so that we can use it on line 94. I mostly referred to https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch, but altered some of the code a bit to be able to do some custom logic for non-200 response codes.

Does that answer your question?